[test] Flaky dynamic row test on master

I've noticed that one of the dynamic row tests fails on master in Chrome:

https://app.circleci.com/pipelines/github/mui/mui-x/19316/workflows/24f01f12-3263-417f-916e-78a39dd5ab81/jobs/112317
https://app.circleci.com/pipelines/github/mui/mui-x/19321/workflows/77d6fa1f-9c92-40e5-8c01-9b3f22568465/jobs/112346

```
Chrome 87.0.4280.66 (Mac OS 10.15.6) <DataGrid /> - Rows prop: getRowHeight dynamic row height should recalculate the content size when the rows prop changes FAILED
	AssertionError: expected inline style of <div …(2)>…(1)</div> did not match
	Expected:
	{
	  "width": "auto",
	  "height": "101px"
	}
	Actual:
	{
	  "width": "auto",
	  "height": "52px"
	}
```

I was able to reproduce it locally:

1. Apply following changes:
	  ```diff
	  --- a/packages/grid/x-data-grid/src/tests/rows.DataGrid.test.tsx
	  +++ b/packages/grid/x-data-grid/src/tests/rows.DataGrid.test.tsx
	  @@ -597,7 +597,7 @@ describe('<DataGrid /> - Rows', () => {
	           });
	         });
	   
	  -      it('should recalculate the content size when the rows prop changes', async () => {
	  +      it.only('should recalculate the content size when the rows prop changes', async () => {
	           const { setProps } = render(
	             <TestCase
	               getBioContentHeight={(row) => (row.expanded ? 200 : 100)}
	  diff --git a/test/karma.conf.js b/test/karma.conf.js
	  index 7160aacbba..4ec2a7f6bf 100644
	  --- a/test/karma.conf.js
	  +++ b/test/karma.conf.js
	  @@ -119,7 +119,7 @@ module.exports = function setKarmaConfig(config) {
	       },
	       customLaunchers: {
	         chromeHeadless: {
	  -        base: 'ChromeHeadless',
	  +        base: 'Chrome',
	           flags: ['--no-sandbox'],
	         },
	       },
	  ```
	It doesn't reproduce in `ChromeHeadless`
	
2. Run tests with `yarn test:karma`
3. On the first run test usually passes without errors.
4. Save any test file without any changes to trigger tests relaunch. Usually subsequent test runs will fail

Not sure why this happens, I tried to debug it but no luck so far

--- a/packages/grid/x-data-grid/src/tests/rows.DataGrid.test.tsx
+++ b/packages/grid/x-data-grid/src/tests/rows.DataGrid.test.tsx
@@ -20,8 +20,6 @@ import { COMPACT_DENSITY_FACTOR } from '../hooks/features/density/useGridDensity
 
 const isJSDOM = /jsdom/.test(window.navigator.userAgent);
 
-const nativeSetTimeout = setTimeout;
-
 describe('<DataGrid /> - Rows', () => {
   const { render, clock } = createRenderer({ clock: 'fake' });
 
@@ -532,10 +530,11 @@ describe('<DataGrid /> - Rows', () => {
           '.MuiDataGrid-virtualScrollerContent',
         );
         const expectedHeight = baselineProps.rows.length * (contentHeight + border);
-        await new Promise((resolve) => nativeSetTimeout(resolve)); // Wait for ResizeObserver to send dimensions
-        expect(virtualScrollerContent).toHaveInlineStyle({
-          width: 'auto',
-          height: `${expectedHeight}px`,
+        await waitFor(() => {
+          expect(virtualScrollerContent).toHaveInlineStyle({
+            width: 'auto',
+            height: `${expectedHeight}px`,
+          });
         });
 
         window.onerror = originalErrorHandler;
@@ -562,10 +561,11 @@ describe('<DataGrid /> - Rows', () => {
           measuredRowHeight +
           border + // Measured rows also include the border
           (baselineProps.rows.length - 1) * defaultRowHeight;
-        await new Promise((resolve) => nativeSetTimeout(resolve)); // Wait for ResizeObserver to send dimensions
-        expect(virtualScrollerContent).toHaveInlineStyle({
-          width: 'auto',
-          height: `${expectedHeight}px`,
+        await waitFor(() => {
+          expect(virtualScrollerContent).toHaveInlineStyle({
+            width: 'auto',
+            height: `${expectedHeight}px`,
+          });
         });
       });
 
@@ -590,10 +590,11 @@ describe('<DataGrid /> - Rows', () => {
         const firstRowHeight = measuredRowHeight + border; // Measured rows also include the border
         const expectedHeight =
           firstRowHeight + (baselineProps.rows.length - 1) * estimatedRowHeight;
-        await new Promise((resolve) => nativeSetTimeout(resolve)); // Wait for ResizeObserver to send dimensions
-        expect(virtualScrollerContent).toHaveInlineStyle({
-          width: 'auto',
-          height: `${expectedHeight}px`,
+        await waitFor(() => {
+          expect(virtualScrollerContent).toHaveInlineStyle({
+            width: 'auto',
+            height: `${expectedHeight}px`,
+          });
         });
       });
 
@@ -609,16 +610,18 @@ describe('<DataGrid /> - Rows', () => {
         const virtualScrollerContent = document.querySelector(
           '.MuiDataGrid-virtualScrollerContent',
         );
-        await new Promise((resolve) => nativeSetTimeout(resolve)); // Wait for ResizeObserver to send dimensions
-        expect(virtualScrollerContent).toHaveInlineStyle({
-          width: 'auto',
-          height: '101px',
+        await waitFor(() => {
+          expect(virtualScrollerContent).toHaveInlineStyle({
+            width: 'auto',
+            height: '101px',
+          });
         });
         setProps({ rows: [{ clientId: 'c1', expanded: true }] }); // Wait for ResizeObserver to send dimensions
-        await new Promise((resolve) => nativeSetTimeout(resolve));
-        expect(virtualScrollerContent).toHaveInlineStyle({
-          width: 'auto',
-          height: '201px',
+        await waitFor(() => {
+          expect(virtualScrollerContent).toHaveInlineStyle({
+            width: 'auto',
+            height: '201px',
+          });
         });
       });
 

